[BUG] Cannot configure SSH git

**Describe the bug**
When configuring git with SSH I enter the following information (personal information redacted):
* Username: `git`
* Server URL: `ssh.example.com`
* Port: `EMPTY`
* Repo path: `:passwords.git` OR `passwords.git` (both fail)
* Authentication Mode: SSH or OpenKeyChain. I didn't attempt password since I don't want to allow that in my repository.

When I press "Clone" it returns an error like
```
Error occurred during the clone operation, Message from jgit:
Invalid remote: origin
Caused by: org.eclipse.jgit.errors.NoRemoteRepositoryException: ssh://git@ssh.example.com/:passwords.git: fatal: '/:passwords.git' does not appear to be a git repository
```

You will notice that it adds a `/` at the start of the repo path, probably from here: https://github.com/android-password-store/Android-Password-Store/blob/041cf00510a75042223d1857660fc7f1dff958ef/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt#L98

**To Reproduce**
Steps to reproduce the behavior:
1. Enter the data I did above.
2. Press "Clone"
3. See the error

**Expected behavior**
The repository should be successfully cloned.

**Device information (please complete the following information):**
 - Device: OnePlus 8 Pro
 - OS: Oxygen OS 10.5.6.IN11AA
 - App version: 1.7.2

**Additional context**
This worked with the older version of the application, but I am setting this up on a new phone and encountered this today.

Additionally, this happens with both an SSH key generated in-app and copied on to the server as well as with an OpenKeyChain GPG authentication key.

Finally, I tested this on my machine with `git clone git@ssh.example.com:passwords.git` and it successfully clones the repository.

--- a/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
+++ b/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
@@ -23,7 +23,6 @@ import com.zeapo.pwdstore.git.config.SshApiSessionFactory
 import com.zeapo.pwdstore.utils.PasswordRepository
 import com.zeapo.pwdstore.utils.getEncryptedPrefs
 import java.io.File
-import java.net.MalformedURLException
 import java.net.URI
 
 /**
@@ -94,34 +93,41 @@ abstract class BaseGitActivity : AppCompatActivity() {
             return false
 
         val previousUrl = url ?: ""
-        val hostnamePart = serverHostname
-        val pathPart = if (serverPath.startsWith('/')) serverPath else "/$serverPath"
+        // Whether we need the leading ssh:// depends on the use of a custom port.
+        val hostnamePart = serverHostname.removePrefix("ssh://")
+        // We only support relative paths and trim everything scheme-specific.
+        val pathPart = serverPath.trimStart('/', ':')
         val newUrl = when (protocol) {
             Protocol.Ssh -> {
-                val userPart = if (serverUser.isEmpty()) "" else "$serverUser@"
+                val userPart = if (serverUser.isEmpty()) "" else "${serverUser.trimEnd('@')}@"
                 val portPart =
                     if (serverPort == "22" || serverPort.isEmpty()) "" else ":$serverPort"
-                if (hostnamePart.startsWith("ssh://"))
-                    hostnamePart.replace("ssh://", "")
-                // We have to specify the ssh scheme as this is the only way to pass a custom port.
-                "ssh://$userPart$hostnamePart$portPart$pathPart"
+                if (portPart.isEmpty()) {
+                    "$userPart$hostnamePart:$pathPart"
+                } else {
+                    // We have to specify the ssh scheme as this is the only way to pass a custom
+                    // port.
+                    "ssh://$userPart$hostnamePart$portPart/$pathPart"
+                }
             }
             Protocol.Https -> {
                 val portPart =
                     if (serverPort == "443" || serverPort.isEmpty()) "" else ":$serverPort"
-                val urlWithFreeEntryScheme = "$hostnamePart$portPart$pathPart"
-                when {
+                val urlWithFreeEntryScheme = "$hostnamePart$portPart/$pathPart"
+                val url = when {
                     urlWithFreeEntryScheme.startsWith("https://") -> urlWithFreeEntryScheme
                     urlWithFreeEntryScheme.startsWith("http://") -> urlWithFreeEntryScheme.replaceFirst("http", "https")
                     else -> "https://$urlWithFreeEntryScheme"
                 }
+                try {
+                    if (URI(url).rawAuthority != null)
+                        url
+                    else
+                        return false
+                } catch (_: Exception) {
+                    return false
+                }
             }
-        }
-        try {
-            if (URI(newUrl).rawAuthority == null)
-                return false
-        } catch (_: MalformedURLException) {
-            return false
         }
         if (PasswordRepository.isInitialized)
             PasswordRepository.addRemote("origin", newUrl, true)

--- a/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
+++ b/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
@@ -17,6 +17,7 @@ import androidx.preference.PreferenceManager
 import com.github.ajalt.timberkt.Timber.tag
 import com.github.ajalt.timberkt.e
 import com.google.android.material.dialog.MaterialAlertDialogBuilder
+import com.zeapo.pwdstore.R
 import com.zeapo.pwdstore.git.config.ConnectionMode
 import com.zeapo.pwdstore.git.config.Protocol
 import com.zeapo.pwdstore.git.config.SshApiSessionFactory
@@ -83,14 +84,25 @@ abstract class BaseGitActivity : AppCompatActivity() {
         super.onDestroy()
     }
 
+    enum class GitUpdateUrlResult(val textRes: Int) {
+        Ok(0),
+        CustomPortRequiresAbsoluteUrlError(R.string.git_config_error_custom_port_absolute),
+        EmptyHostnameError(R.string.git_config_error_hostname_empty),
+        GenericError(R.string.git_config_error_generic),
+        NonNumericPortError(R.string.git_config_error_nonnumeric_port)
+    }
+
     /**
-     * Update the [url] field with the values that build it up. This function returns a boolean
-     * indicating whether or not the values are likely valid or not, and only adds the `origin`
-     * remote when it is. This check is not perfect, it is mostly meant to catch typos.
+     * Update the [url] field with the values that build it up. This function returns a
+     * [GitUpdateUrlResult] indicating whether the values could be used to build a URL and only adds
+     * the `origin` remote when they were. This check is not perfect, it is mostly meant to catch
+     * syntax-related typos.
      */
-    fun updateUrl(): Boolean {
-        if (serverHostname.isEmpty() || !serverPort.isDigitsOnly())
-            return false
+    fun updateUrl(): GitUpdateUrlResult {
+        if (serverHostname.isEmpty())
+            return GitUpdateUrlResult.EmptyHostnameError
+        if (!serverPort.isDigitsOnly())
+            return GitUpdateUrlResult.NonNumericPortError
 
         val previousUrl = url ?: ""
         // Whether we need the leading ssh:// depends on the use of a custom port.
@@ -105,9 +117,9 @@ abstract class BaseGitActivity : AppCompatActivity() {
                     val pathPart = serverPath.trimStart('/', ':')
                     "$userPart$hostnamePart:$pathPart"
                 } else {
-                    // We only support absolute paths with custom ports.
+                    // Only absolute paths are supported with custom ports.
                     if (!serverPath.startsWith('/'))
-                        return false
+                        return GitUpdateUrlResult.CustomPortRequiresAbsoluteUrlError
                     val pathPart = serverPath
                     // We have to specify the ssh scheme as this is the only way to pass a custom
                     // port.
@@ -128,9 +140,9 @@ abstract class BaseGitActivity : AppCompatActivity() {
                     if (URI(url).rawAuthority != null)
                         url
                     else
-                        return false
+                        return GitUpdateUrlResult.GenericError
                 } catch (_: Exception) {
-                    return false
+                    return GitUpdateUrlResult.GenericError
                 }
             }
         }
@@ -141,7 +153,7 @@ abstract class BaseGitActivity : AppCompatActivity() {
         if (previousUrl.isNotEmpty() && newUrl != previousUrl && protocol == Protocol.Https)
             encryptedSettings.edit { remove("https_password") }
         url = newUrl
-        return true
+        return GitUpdateUrlResult.Ok
     }
 
     /**

--- a/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
+++ b/app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
@@ -95,24 +95,29 @@ abstract class BaseGitActivity : AppCompatActivity() {
         val previousUrl = url ?: ""
         // Whether we need the leading ssh:// depends on the use of a custom port.
         val hostnamePart = serverHostname.removePrefix("ssh://")
-        // We only support relative paths and trim everything scheme-specific.
-        val pathPart = serverPath.trimStart('/', ':')
         val newUrl = when (protocol) {
             Protocol.Ssh -> {
                 val userPart = if (serverUser.isEmpty()) "" else "${serverUser.trimEnd('@')}@"
                 val portPart =
                     if (serverPort == "22" || serverPort.isEmpty()) "" else ":$serverPort"
                 if (portPart.isEmpty()) {
+                    // We only support relative paths with the standard port.
+                    val pathPart = serverPath.trimStart('/', ':')
                     "$userPart$hostnamePart:$pathPart"
                 } else {
+                    // We only support absolute paths with custom ports.
+                    if (!serverPath.startsWith('/'))
+                        return false
+                    val pathPart = serverPath
                     // We have to specify the ssh scheme as this is the only way to pass a custom
                     // port.
-                    "ssh://$userPart$hostnamePart$portPart/$pathPart"
+                    "ssh://$userPart$hostnamePart$portPart$pathPart"
                 }
             }
             Protocol.Https -> {
                 val portPart =
                     if (serverPort == "443" || serverPort.isEmpty()) "" else ":$serverPort"
+                val pathPart = serverPath.trimStart('/', ':')
                 val urlWithFreeEntryScheme = "$hostnamePart$portPart/$pathPart"
                 val url = when {
                     urlWithFreeEntryScheme.startsWith("https://") -> urlWithFreeEntryScheme

--- a/app/src/main/java/com/zeapo/pwdstore/git/GitServerConfigActivity.kt
+++ b/app/src/main/java/com/zeapo/pwdstore/git/GitServerConfigActivity.kt
@@ -97,22 +97,23 @@ class GitServerConfigActivity : BaseGitActivity() {
         binding.saveButton.setOnClickListener {
             if (isClone && PasswordRepository.getRepository(null) == null)
                 PasswordRepository.initialize(this)
-            if (updateUrl()) {
-                settings.edit {
-                    putString("git_remote_protocol", protocol.pref)
-                    putString("git_remote_auth", connectionMode.pref)
-                    putString("git_remote_server", serverHostname)
-                    putString("git_remote_port", serverPort)
-                    putString("git_remote_username", serverUser)
-                    putString("git_remote_location", serverPath)
+            when (val result = updateUrl()) {
+                GitUpdateUrlResult.Ok -> {
+                    settings.edit {
+                        putString("git_remote_protocol", protocol.pref)
+                        putString("git_remote_auth", connectionMode.pref)
+                        putString("git_remote_server", serverHostname)
+                        putString("git_remote_port", serverPort)
+                        putString("git_remote_username", serverUser)
+                        putString("git_remote_location", serverPath)
+                    }
+                    if (!isClone) {
+                        Snackbar.make(binding.root, getString(R.string.git_server_config_save_success), Snackbar.LENGTH_SHORT).show()
+                        Handler().postDelayed(500) { finish() }
+                    } else
+                        cloneRepository()
                 }
-                if (!isClone) {
-                    Snackbar.make(binding.root, getString(R.string.git_server_config_save_success), Snackbar.LENGTH_SHORT).show()
-                    Handler().postDelayed(500) { finish() }
-                } else
-                    cloneRepository()
-            } else {
-                Snackbar.make(binding.root, getString(R.string.git_server_config_save_failure), Snackbar.LENGTH_LONG).show()
+                else -> Snackbar.make(binding.root, getString(R.string.git_server_config_save_error_prefix, getString(result.textRes)), Snackbar.LENGTH_LONG).show()
             }
         }
     }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -363,7 +363,11 @@
     <string name="connection_mode_openkeychain" translatable="false">OpenKeychain</string>
     <string name="connection_mode_none">None</string>
     <string name="git_server_config_save_success">Successfully saved configuration</string>
-    <string name="git_server_config_save_failure">Configuration error: please verify your settings and try again</string>
+    <string name="git_server_config_save_error_prefix">Configuration error: %s</string>
+    <string name="git_config_error_hostname_empty">empty hostname</string>
+    <string name="git_config_error_generic">please verify your settings and try again</string>
+    <string name="git_config_error_nonnumeric_port">port must be numeric</string>
+    <string name="git_config_error_custom_port_absolute">path must be absolute (start with \'/\') when using a custom port</string>
     <string name="git_operation_unable_to_open_ssh_key_title">Unable to open the ssh-key</string>
     <string name="git_operation_unable_to_open_ssh_key_message">Please check that it was imported.</string>
     <string name="git_operation_wrong_passphrase">Wrong passphrase</string>

